### PR TITLE
add map, list, set hooks in rust

### PR DESCRIPTION
**Pull request by dwightguth**
@bmmoore your review of the rust code would be appreciated. @theo25 feel free to review the function signatures and changes to header.ll as well as the test coverage and any rust you feel you can understand.

### CODE REVIEW

File: runtime/datastructures/src/hook_map.rs [hunk 1]
```diff
@@ -0,0 +1,394 @@
...
+
+#[no_mangle]
+pub unsafe extern "C" fn hook_MAP_remove(result: *mut Map, m: *const Map, key: K) -> bool {
+  *result = (*m).without(&key);
+  true
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn hook_MAP_difference(result: *mut Map, m1: *const Map, m2: *const Map) -> bool {
+  *result = (*m1).clone().difference_with((*m2).clone(), |v1, v2| if v1 == v2 { None } else { Some(v1) });
```
[the remark this answers is not in the document]
Comment (dwightguth) [reply runtime/datastructures/src/hook_map.rs]: So it removes any keys in m1 which are bound to the same value in m2. That is the same behavior as the ocaml backend, and as you can see below, it's being tested. Is there a particular case you think is incorrect?

File: runtime/datastructures/src/hook_map.rs [hunk 2]
```diff
@@ -0,0 +1,394 @@
...
+  Box::into_raw(b)
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn free_map(ptr: *mut Map) {
+  Box::from_raw(ptr);
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn hook_MAP_element(result: *mut Map, key: K, value: K) -> bool {
```
[the remark this answers is not in the document]
Comment (dwightguth) [reply runtime/datastructures/src/hook_map.rs]: The intent of the signature of the hooks is that the K value of the arguments in the llvm backend are passed to the hook as-is, the result is allocated outside the hook itself, and the hook writes to the previously allocated result.

**Comment by dwightguth** [comment]
@bmmoore this is ready for you to look at again.

File: runtime/datastructures/src/hook_map.rs [hunk 3]
```diff
@@ -0,0 +1,404 @@
...
+  if (*m).is_empty() {
+    return false;
+  }
+  ptr::write(result, *(*m).keys().next().unwrap());
+  true
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn hook_MAP_size(result: *mut Int, m: *const Map) -> bool {
+  __gmpz_set_ui(result, (*m).len());
```
Comment (bmmoore) [inline runtime/datastructures/src/hook_map.rs]: Does this require the int to already be initialized? Many of the other methods work with uninitialized targets, and would leak memory if called with result pointing to an initialized value.

File: runtime/datastructures/src/hook_map.rs [hunk 4]
```diff
@@ -0,0 +1,404 @@
...
+  ptr::write(result, (*m2).clone().union((*m1).clone()));
+  true
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn hook_MAP_removeAll(result: *mut Map, map: *const Map, set: *const Set) -> bool {
+  let mut tmp = (*map).clone();
+  for key in (*set).iter() {
+    tmp.remove(key);
+  }
```
Comment (bmmoore) [inline runtime/datastructures/src/hook_map.rs]: I'm surprised there wasn't already a method, but this seems to be a reasonable implementation (the code for `difference` also involves a loop calling `remove`).

File: runtime/datastructures/src/hook_list.rs [hunk 5]
```diff
@@ -0,0 +1,12 @@
+use super::decls::List;
+
+#[no_mangle]
+pub extern "C" fn alloc_list() -> *mut List {
```
Comment (bmmoore) [inline runtime/datastructures/src/hook_list.rs]: List and Set should probably have the allocation style changed as well.

File: runtime/datastructures/src/hook_set.rs [hunk 6]
```diff
@@ -0,0 +1,12 @@
+use super::decls::Set;
+
+#[no_mangle]
+pub extern "C" fn alloc_set() -> *mut Set {
```
Comment (bmmoore) [inline runtime/datastructures/src/hook_set.rs]: List and Set should probably have the allocation style changed as well.

**Comment by bmmoore** [comment]
`hook_MAP_lookupOrDefault` should do the `match` in the same style as the function before it.

File: runtime/datastructures/src/hook_map.rs (hunk 3 above)
Comment (dwightguth) [reply runtime/datastructures/src/hook_map.rs]: the assumption is that some initialization will have already occurred when the heap block is created, and I'm including the call to mpz_init in that category.

File: runtime/datastructures/src/decls.rs [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+pub struct Int(
+  i32, // _mp_alloc
+  i32, // _mp_size
+  *const mp_limb_t, // _mp_d
+);
+
+pub type K = *const Block;
+pub type Map = HashMap<K, K>;
+pub type Set = HashSet<K>;
+pub type List = Vector<K>;
```
Comment (bmmoore) [inline runtime/datastructures/src/decls.rs]: Let's see how it goes, but based on the documentation the minimum size of this Vector is a bit high:
https://docs.rs/im/11.0.1/im/vector/index.html
Perhaps 2 or 3 of those arrays will often be shared for small vectors, and we could add further small-vector optimizations if needed.

**Pull request merged**